# Prelecture

## Question 1 

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

The df.isna().sum() shows that there are missing values in this dataset

## Question 2

### 2.1

In [2]:
df.shape

(391, 11)

The dataset contains 11 columns and 391 rows. 

### 2.2

Definition of Obervations: Observations are the individual records that are stored in a dataset where each of them corresponds to a specific character/row. 

Definition of Variables: Variables correspond to the columns of the dataset where each of them refers to a attribute or property of an observation

## Question 3

In [3]:
df['species'].value_counts()

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [4]:
# Get descriptive statistics for all columns
df.describe(include='all')

row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391               391   
unique            92      388               391   
top     K.K. Country  wee one  villager-admiral   
freq              10        2                 1   
mean             NaN      NaN               NaN   
std              NaN      NaN               NaN   
min              NaN      NaN               NaN   
25%              NaN      NaN               NaN   
50%              NaN      NaN               NaN   
75%              NaN      NaN               NaN   
max              NaN      NaN               NaN   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [5]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [6]:
df.shape

(391, 11)

## Question 4

In [7]:
url2 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df2 = pd.read_csv(url2)
df2.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [8]:
df2.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [9]:
df2.shape

(891, 15)

In [10]:
df2.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [11]:
df2.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

#### The discrepancies between size of the dataset given by df.shape and what is reported by df.describe()
By default, df.describe() method only includes numerical columns as it provides summary statistics like count, mean, standard deviation, etc. This results in fewer columns being displayed compared to the total number of columns in the datase since non-numerical columns were not displayed. We can use df.describe(include='all') method to show all columns regardless of their data type. 

The "count" reported by df.describe only counts the number of non-missing values in column. For example, it counts 714 values in gthe "age" column because there are 177 missing values in that column.

## Question 5

#### Difference between attribute and method:
An attribute is a characteristic or property of an object like the color of a car for example. 

A method is a function that when it's performed, it does some operation or action to the object for example you start the engine of a car. 


### Summary of Interaction for Prelecture Section

Chatlog from Microsoft Copilot: https://copilot.microsoft.com/sl/jztIXw4ShUG

**1. Dataset Information:**
- You downloaded a dataset about characters from Animal Crossing from [this URL](https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv).
- The dataset contains 391 rows and 12 columns, including information such as name, species, gender, personality, birthday, catchphrase, hobby, saying, style preferences, and color preferences.

**2. Python Code for Dataset Summary:**
- I provided a Python script to load the dataset and display a summary using `df.describe(include='all')` and count unique values in a specific column using `df['column'].value_counts()`.

**3. Explanation of Terms:**
- **Observations**: Individual entries or rows in the dataset (391 villagers).
- **Variables**: Columns in the dataset representing different attributes (12 attributes).

**4. Additional Dataset:**
- You provided another dataset URL for the Titanic dataset: [this URL](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv).
- I explained the columns in the Titanic dataset and provided a Python script to load the dataset, check for missing values, display a summary, and count unique values in the 'sex' column.

**5. Difference Between Attributes and Methods:**
- **Attributes**: Characteristics or properties of an object, accessed without parentheses (e.g., `df.shape`).
- **Methods**: Functions associated with an object, performing actions, and called with parentheses (e.g., `df.describe()`).

---


# Postlecture

## Question 6

- Count: The number of existing values in each column
- Mean: The average of values in each column
- Std: The standard deviation of the values in each column, this measures the amount of variation from the mean
- Min: The minimum value in each column
- 25%: The 25th percentile(1st Quartile) is the value which 25% of the data lies below it
- 50%: The 50th percentile(median) is the value which 50% of the data lies below it
- 75%: The 75th percential(3rd Quartile) is the value which 75% of the data lies below it
- Max: The maximum value in each column

When there are missing values in the columns that df.describe() analyzes, it ignores and computes using only the non-missing values. 

## Question 7

### 7.1

### df.dropna() is Preferred over del df['col'] in the Following Scenarios
- A small portion of the data is missing a column, so dropping rows is more efficient than removing the entire columns.
- When there's multiple missing values in a row, dropping it is more practical than deleting columns.
- If you plan to use the column in other analysis, it's better to drop rows rather than deleting the entire column.

Below is an example:

In [12]:
# Drop rows with missing 'age' values, but keep the 'age' column for future analysis
df2.dropna(subset=['age'])

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
885  woman       False  NaN   Queenstown    no  False  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[714 rows x 15 columns]

### 7.2

### del df['col'] is Preferred over df.dropna() in the Following Scenarios
- When a column has a significant portion of it's data missing, discard the entire column is more efficient since keeping the column may not add much value to the analysis.
- When a column is not relevent to our analysis and any other analysis we plan to do, we can just remove it to optimize our memory.
- When a column contains only a single value (or nearly the same value across all rows). It doesn't provide any useful information for our analysis. 

Below is an example:

In [13]:
# Drop the 'deck' column since it's mostly missing and not useful
del df2['deck']
df2

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[891 rows x 14 columns]

### 7.3

### Why Applying del df['col'] before df.dropna()
- If there is a column that has too much missing values when df.dropna() is applied, it could cause excessive row dropping which means rows with valid data may get removed just because they have missing data in a column that is irrelevant or already slated for deletion.
- By deleting the columns you don't need early on, it reduces the dimensionality of the dataframe which speed up computation.

### 7.4

### Remove All Missing Data from the Dataset using Combination of del df['col'] and/or df.dropna()

In [14]:
df3 = pd.read_csv(url2)
df3

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [15]:
# Calculate the percentage of missing values per column
df3.isna().mean()*100

survived        0.000000
pclass          0.000000
sex             0.000000
age            19.865320
sibsp           0.000000
parch           0.000000
fare            0.000000
embarked        0.224467
class           0.000000
who             0.000000
adult_male      0.000000
deck           77.216611
embark_town     0.224467
alive           0.000000
alone           0.000000
dtype: float64

In [16]:
# Remove the column that has over 50% of it's data missing
del df3['deck']
df3

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[891 rows x 14 columns]

In [17]:
# Now, we can remove any rows that still contain missing values in the remaining columns.
df3.dropna()

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
885  woman       False   Queenstown    no  False  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[712 rows x 14 columns]

### Before and After Report
The way that I removed all the missing values is that I first analyzes the missing data by calculating percentage of missing data in each column. Then I removed the columns that has over 50% of it's data missing. Finally, I removed the rows with missing data. The dimension of the dataset was 891×15 before and 712×14 after. I think this method of removing the data ensure's the data integrity to the maximum extent possible. 

### Summary of Interaction for Q6 & Q7: 

Transcript Log: https://chatgpt.com/share/60d02642-cb30-413b-ad2b-d05bdedad148

#### Objective:
The aim of this session was to explore how to handle missing data efficiently in a dataset using the Titanic dataset from the seaborn repository, with a focus on maximizing the use of available data.

#### Key Concepts Covered:

1. **Understanding `df.describe()`**:
   - We reviewed the meaning of `count`, `mean`, `std`, `min`, `25%`, `50%`, `75%`, and `max`, as provided by `df.describe()`, which summarize the central tendency and spread of numerical data, ignoring missing values.

2. **Handling Missing Data**:
   - **`df.dropna()`**: This method was discussed for removing rows or columns that contain missing values. It drops rows by default but can also drop columns when specified with `axis=1`.
   - **`del df['col']`**: We examined how this approach directly removes columns from a DataFrame and is useful when certain columns are irrelevant or contain too many missing values to be useful in the analysis.

3. **Combining `del df['col']` with `df.dropna()`**:
   - We discussed why applying `del df['col']` before `df.dropna()` can be critical when aiming to remove columns with excessive missing values to prevent the loss of too many rows.
   - Using `del` ensures that irrelevant columns with a large amount of missing data don't cause unnecessary row deletions when applying `df.dropna()`.

4. **Scenario for Using `del df['col']` vs `df.dropna()`**:
   - **Use `del df['col']`**: When a column is irrelevant or has a significant amount of missing data and removing it reduces the risk of deleting important rows.
   - **Use `df.dropna()`**: To remove rows with missing values from the remaining dataset, ensuring the data is complete after removing unnecessary columns.

5. **Practical Application**:
   - We applied the above concepts to the Titanic dataset by first removing irrelevant columns (`deck`, `cabin`, etc.) with too many missing values using `del df['col']`, and then using `df.dropna()` to remove any rows that still contained missing data.

6. **Maximizing Data Retention**:
   - The focus was on keeping as much data as possible, including irrelevant columns, while ensuring that no missing data remains. We set a threshold to remove columns with a high percentage of missing values (e.g., > 50%) before using `df.dropna()` to clean the remaining rows.
   - Example:
     ```python
     del df['deck']
     del df['cabin']
     df_cleaned = df.dropna()
     ```

#### Conclusion:
This interaction provided a detailed strategy for efficiently handling missing data by combining the use of `del df['col']` for irrelevant or problematic columns and `df.dropna()` for rows with missing values. This approach ensures that valuable data is retained, while irrelevant columns or excessive missing data are removed in a targeted way.

This method is particularly useful when working with datasets that contain both critical and less important columns, allowing for data cleaning without excessive loss of rows or columns that might still hold useful information.

This summary reflects the overall workflow and thought process used during this session to clean and prepare the dataset effectively for further analysis.

## Question 8

In [18]:
df4 = pd.read_csv(url2)
df4

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 8.1

### Function of df.groupby("col1")["col2"].describe() 
- First, it splits the data into subsets/groups where each subset/group contains all rows that share the same value in "col1".
- Next, it specifies that the operation should be applied to column "col2" for each group from the previous step.
- Finally, .describe() calculates the statistics (like count, mean, std, min, quartiles, and max) for "col2" within each group created by "col1".

Below is an example from the "titanic" data set:

In [19]:
df4.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [20]:
df4.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

### 8.2

### The Different Values of Count Results from df.describe() and df.groupby("col1")["col2"].describe()
- The count results from df.describe() is the number of non-missing values in that column. 
- The count results from df.groupby("col1")["col2"].describe() shows the number of non-missing values in "col2" for each group in "col1". 

This can be shown from the above example that the sum of count of male and female is 714 which is not the count for 'sex' from df.describe, but rather the count for 'age'. 

Therefore, the function of count in conjunction with group_by is to show the number of non-missing values in "col2" for each group in "col1". 

### Summary of Interaction for Q8.1 & Q8.2:

Transcript Log: https://chatgpt.com/share/d446a148-7d05-465d-8684-4f79fccee851

In this session, I performed an initial analysis on the Titanic dataset downloaded from the seaborn repository. I explored functions related to grouping and summarizing data using pandas, specifically focusing on handling missing values and descriptive statistics.

I used `df.groupby("col1")["col2"].describe()` to understand how pandas groups data by one column and then provides descriptive statistics for another column within each group. I learned that:
- `groupby()` splits the DataFrame into subsets based on unique values in the specified column.
- `describe()` computes summary statistics (like count, mean, standard deviation, etc.) for each group.

I also explored how the **count** in pandas' `describe()` function behaves:
- The **count** within `describe()` reflects the number of **non-NaN values** in the grouped column, not the total number of rows in each group.
- When there are missing values (NaNs), the count will exclude those, resulting in fewer entries being counted.

I discussed the difference between using `df.describe()` and `df.groupby("col1")["col2"].describe()`:
- `df.describe()` gives a summary for the whole dataset, while `groupby()` allows us to get group-specific statistics.
- Additionally, the primary function of `count` in `groupby` is to calculate the frequency of non-NaN values within each group.

This understanding of grouping, counting, and handling missing values is essential for analyzing data and preparing it for further modeling or insights.


### 8.3A

#### NameError: name 'pd' is not defined
For this error, chatGPT is definitely faster because it gives the solution right away. Google on the other hand, requires me to click into another site and scroll down to find the answer. 

### 8.3B

#### FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'
For this error, Google is faster as it gives the answer right away by telling me "there is no file of that name in the working directory". ChatGPT on the other hand gives me a list of possible solution which I need to check one by one. 

### 8.3C

#### NameError: name 'df5' is not defined
For this error, both Google and ChatGPT provides good solution by saying that the DataFrame is not defined, but I will say that ChatGPT is more convenience since it provides me with the code to create this variable. 

### 8.3D

#### SyntaxError: incomplete input
For this error, ChatGPT is faster because the first solution it spits out is "check all parentheses are properly closed". With Google however, I need to do a bit of research to find the solution. 

### 8.3E

#### AttributeError: 'DataFrame' object has no attribute 'group_by'
Again, both Google and ChatGPT provides the correct solution for this error, and again ChatGPT provides the correct code, so I think ChatGPT wins this round. 

### 8.3F

#### KeyError: 'Sex'
For this error, ChatGPT is faster because it gives the solution immediately, and Google on the other hand, requires me to do some scrolling in order to find the solution. 

### 8.3G

#### NameError: name 'age' is not defined
For this error, neither ChatGPT nor Google spits the correct solution. However, I think ChatGPT is more useful since it spits out some code that has the column 'age' in quotation marks. 

### Interaction Summary for Q8.3:

Transcript Log: https://chatgpt.com/share/415f6508-d7e7-4487-87bd-16ce014f202f

1. **Fixing `NameError: name 'pd' is not defined`:**
   - Solution: Add `import pandas as pd` at the beginning of the script to ensure pandas is imported correctly.

2. **Fixing `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'`:**
   - Solution: Check the file name and location, ensure it's in the correct directory, and provide the correct file path (either full or relative).

3. **Fixing `NameError: name 'df5' is not defined`:**
   - Solution: Ensure that `df5` is defined and assigned earlier in the code and verify the order of code execution to avoid referencing an undefined variable.

4. **Fixing `SyntaxError: incomplete input`:**
   - Solution: Look for unclosed parentheses, missing colons, incomplete code blocks, and other typical code completion errors (like missing quotes in strings or using multiline statements correctly).

5. **Fixing `AttributeError: 'DataFrame' object has no attribute 'group_by'`:**
   - Solution: The correct method is `groupby`, not `group_by`. Replace `.group_by` with `.groupby` to fix the error.

6. **Fixing `KeyError: 'Sex'`:**
   - Solution: Ensure the column name is correct (case-sensitive) and check for hidden whitespace. You can print the column names and strip any extra spaces if necessary.

7. **Fixing `NameError: name 'age' is not defined`:**
   - Solution: Make sure the variable `age` is defined before using it. If `age` is a column in a DataFrame, reference it correctly using `df['age']`.



## Question 9
# YES